elastic/kibana
# [ResponseOps][Stack Connectors] Opsgenie backend

This PR implements the backend of the opsgenie connector.

Issue: https://github.com/elastic/kibana/issues/142776

Opsgenie API Docs: https://docs.opsgenie.com/docs/alert-api#create-alert

## Examples


### Creating alerts

<details><summary>Postman creating connector</summary>

<img width="614" alt="create_connector" src="https://user-images.githubusercontent.com/56361221/193914479-5cdc3855-9d81-4144-873e-0d361bd70501.png">


</details>



<details><summary>Postman creating an alert</summary>

<img width="757" alt="creating_alert" src="https://user-images.githubusercontent.com/56361221/193914538-1cb4687d-44d3-48c9-a5f0-a6a22b576e77.png">

</details>




<details><summary>Created Alerts within Opsgenie</summary>

<img width="1320" alt="image" src="https://user-images.githubusercontent.com/56361221/193914276-d9d28816-5b7b-408a-95e3-c6d7085aa5d9.png">


</details>

### Closing alerts

<details><summary>Postman closing an alert</summary>

<img width="631" alt="image" src="https://user-images.githubusercontent.com/56361221/193922235-e7e25ab4-51e2-438c-bc77-f3e9f28621cb.png">


</details>

<details><summary>Closed alert within Opsgenie</summary>

<img width="1254" alt="image" src="https://user-images.githubusercontent.com/56361221/193922140-6d692717-e97f-4c9f-8e36-bab763578caf.png">


</details>

## Testing

1. Create an Opsgenie account here: https://www.atlassian.com/software/opsgenie
2. Create a team
3. On the team page there should be `Integrations` on the left, click that, then click `Add Integration`
4. Search for and add `API Integration`
5. After the integration is added, it will create an API key for use in the following APIs, save this key for later
6. Click `Save Integration`

### Creating a connector

<details><summary>curl</summary>

```
curl --location --request POST 'http://elastic:changeme@localhost:5601/api/actions/connector' \
--header 'kbn-xsrf: foo' \
--header 'Content-Type: application/json' \
--data-raw '{
    "name": "Opsgenie",
    "connector_type_id": ".opsgenie",
    "config": {
        "apiUrl": "https://api.opsgenie.com"
    },
    "secrets": {
        "apiKey": "<apiKey>"
    }
}'
```

</details>

<details><summary>Raw</summary>

```
POST http://elastic:changeme@localhost:5601/api/actions/connector

{
    "name": "Opsgenie",
    "connector_type_id": ".opsgenie",
    "config": {
        "apiUrl": "https://api.opsgenie.com"
    },
    "secrets": {
        "apiKey": "<apiKey>"
    }
}
```

</details>


### Closing an alert

<details><summary>curl</summary>

```
curl --location --request POST 'http://elastic:changeme@localhost:5601/api/actions/connector/<connector_id>/_execute' \
--header 'kbn-xsrf: foo' \
--header 'Content-Type: application/json' \
--data-raw '{
    "params": {
        "subAction": "createAlert",
        "subActionParams": {
            "message": "It worked"
        }
    }
}'
```

</details>

<details><summary>Raw</summary>

```
POST http://elastic:changeme@localhost:5601/api/actions/connector/<connector_id>/_execute

{
    "params": {
        "subAction": "closeAlert",
        "subActionParams": {
            "alias": "123"
        }
    }
}
```

</details>
